examples/scripts/interactive/compvis-viewer.js

```diff
@@ -0,0 +1,603 @@
+  // switch for second gradient i.e white to red for heat map
+  if (percent >= 1) {
+    percent -= 1
+    startColor = endColor
+    endColor = thirdColor
+  }
+
+  // get colors
+  var startRed = parseInt(startColor.substr(0, 2), 16)
+  var startGreen = parseInt(startColor.substr(2, 2), 16)
+  var startBlue = parseInt(startColor.substr(4, 2), 16)
+
+  var endRed = parseInt(endColor.substr(0, 2), 16)
+  var endGreen = parseInt(endColor.substr(2, 2), 16)
+  var endBlue = parseInt(endColor.substr(4, 2), 16)
+
+  // calculate new color
+  var diffRed = endRed - startRed
+  var diffGreen = endGreen - startGreen
+  var diffBlue = endBlue - startBlue
+
+  diffRed = ((diffRed * percent) + startRed).toString(16).split('.')[0]
+  diffGreen = ((diffGreen * percent) + startGreen).toString(16).split('.')[0]
+  diffBlue = ((diffBlue * percent) + startBlue).toString(16).split('.')[0]
+
+  // ensure 2 digits by color (necessary)
+  if (diffRed.length === 1) diffRed = '0' + diffRed
+  if (diffGreen.length === 1) diffGreen = '0' + diffGreen
+  if (diffBlue.length === 1) diffBlue = '0' + diffBlue
+
+  return '0x' + diffRed + diffGreen + diffBlue
+}
+
+function makeGradientArray () {
+  var gradientArray = []
+  for (var count = 0; count < 101; count++) {
+    var newColor = getGradientColor('FF0000', 'FFFFFF', '0000FF', (0.02 * count))
+    var numColor = parseInt(Number(newColor), 10)
+    gradientArray.push(numColor)
+  }
+  return gradientArray
+}
+
+var gradientArray = makeGradientArray()
+
+// var ligandSele = '( not polymer or not ( protein or nucleic ) ) and not ( water or ACE or NH2 )'
+
+var pocketRadius = 0
+var pocketRadiusClipFactor = 1
+
+var cartoonRepr, neighborRepr, ligandRepr, contactRepr, pocketRepr, labelRepr, customRepr
+
+var heatMap, customPercent
+
+var struc
+var csv
+var neighborSele
+var sidechainAttached = false
+
+const csvResNumCol = 4
+const csvWtProbCol = 7
+const csvPrAaCol = 6
+const csvPrProbCol = 8
+
+var firstResNum
+
+function loadStructure (proteinFile, csvFile) {
+  struc = undefined
+  stage.setFocus(0)
+  stage.removeAllComponents()
+  clipNearRange.value = 0
+  clipRadiusRange.value = 100
+  pocketOpacityRange.value = 0
+  cartoonCheckbox.checked = true
+  customCheckbox.checked = false
+  hydrophobicCheckbox.checked = false
+  hydrogenBondCheckbox.checked = true
+  weakHydrogenBondCheckbox.checked = false
+  waterHydrogenBondCheckbox.checked = true
+  backboneHydrogenBondCheckbox.checked = true
+  halogenBondCheckbox.checked = true
+  metalInteractionCheckbox.checked = true
+  saltBridgeCheckbox.checked = true
+  cationPiCheckbox.checked = true
+  piStackingCheckbox.checked = true
+  return Promise.all([
+    stage.loadFile(proteinFile, { defaultRepresentation: true }),
+    NGL.autoLoad(csvFile, {
+      ext: 'csv',
+      delimiter: ' ',
+      comment: '#',
+      columnNames: true
+    })
+  ]).then(function (ol) {
+    struc = ol[0]
+    csv = ol[1].data
+
```

Comment:
Rather than iterating through csv in the `atomColor` function below, might it be better to build a lookup object here

```javascript
var residueData = {}
for (var i = 0; i < csv.length; i++) {
    resNum = parseFloat(csv[i][csvResNumCol]
    residueData[resNum] = csv[i]
}
```

Reply:
Looking at the other PDB you pushed (6ar1) you've got data for 2 chains (so the residue numbers restart). I'd update this to use the combination of chain name and residue number to index `residueData`, e.g.:

```javascript
var residueData = {}   // Ends up as { 'A101': csv[x], 'A102': csv[y], ...}
for (var i = 0; i < csv.length; i++) {
    chain = csv[i][chainCol] // chainCol is 3 in your example, `chain` will be something like 'A'
    resNum = parseFloat(csv[i][csvResNumCol] // e.g. 101
    residueData[chain+resNum] = csv[i]  // chain+resNum will be 'A101'
}
```